### PR TITLE
Fixes and improvements

### PR DESCRIPTION
**Pull request by apaszke**
* Added better error message for CUDA -> numpy conversions (#779)
* Added checks for type and size of rewards in StochasticFunction (#781)
* cuDNN RNN Function now returns a correct number of gradients (#776)
* input will be made contiguous only once in backward of cuDNN RNN (#621)
* DataParallel now supports multiple inputs (#649)
* Added `torch.nn.utils.clip_grad_norm` (#309)
* Added support for indexing with `None` and slices with positive steps (partially #229)
* Added support for numpy arrays in `default_collate` (#782)
* Fixed misspelling and made NLLLoss2d accept weights (#784)

NLLLoss2d test will be extremely flaky until torch/cunn#445 is merged.

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
rebased and merged
